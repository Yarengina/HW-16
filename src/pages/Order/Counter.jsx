// eslint-disable-next-line no-unused-vars
import * as React from 'react'

function Counter({ min, max, current, onChange }) {
    function applyCurrent(num) {
        const validNum = Math.max(min, Math.min(max, num))
        onChange(validNum)
    }

    const inc = () => applyCurrent(current + 1)
    const dec = () => applyCurrent(current - 1)

    return (
        <div className="Counter">
            <button className="Buttons Plus" type="button" onClick={dec}>
                Remove item
            </button>
            <p className="Quantity">{current - 1}</p>
            <button className="Buttons Minus" type="button" onClick={inc}>
                Add to cart
            </button>
        </div>
    )
}

export default Counter
