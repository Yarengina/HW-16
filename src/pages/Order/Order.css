.Total {
    text-align: initial;
}

.BookField {
    display: flex;
    flex-wrap: wrap;
    gap: 20px;
    margin-bottom: 30px;
}

.Book {
    width: 200px;
    height: 200px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    background-color: white;
    border: 1px solid rgb(153, 153, 153);
    border-radius: 4px;
}

.Title {
    width: 150px;
    font-weight: 700;
    margin: 15px;
}

.Price {
    text-align: initial;
    padding-left: 10px;
}

.ReadMoreLink {
    color: #0000ffbd;
    margin-left: 10px;
    margin-bottom: 15px;
}

.Counter {
    display: flex;
    margin-left: 10px;
    margin-right: 10px;
}

.Quantity {
    width: 10px;
    color: gray;
    padding-top: 2px;
    margin: 0px 2px;
}

.Buttons {
    width: 84px;
    height: 30px;
    font-size: 11px;
    font-weight: bold;
    color: white;
    border: initial;
    border-radius: 3px;
}

.Minus {
    background-color: #4caf50;
}

.Plus {
    background-color: #ff0000a6;
}

.ButtonReady {
    width: 100px;
    height: 34px;
    background-color: #4caf50;
    border: initial;
    border-radius: 4px;
    color: white;
    font-weight: bolder;
    margin-left: 6px;
}
