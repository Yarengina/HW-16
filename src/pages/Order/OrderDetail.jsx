// eslint-disable-next-line no-unused-vars
import * as React from 'react'
import { Link } from 'react-router-dom'
import Counter from './Counter'

function OrderDetail({ books, setQuantity }) {
    return (
        <div className="BookField">
            {books.map(({ id, title, price, rest, quantity }) => (
                <div data-testid={`todo-${id}`} className="Book" key={id}>
                    <p className="Title">{title}</p>
                    <p className="Price">{price}</p>
                    <Link className="ReadMoreLink" to={`${id}`}>
                        Read more
                    </Link>
                    <Counter
                        min={1}
                        max={rest + 1}
                        current={quantity}
                        onChange={(bookQuantity) =>
                            setQuantity(id, bookQuantity)
                        }
                    />
                </div>
            ))}
        </div>
    )
}

export default OrderDetail
