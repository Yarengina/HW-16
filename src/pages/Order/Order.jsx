// eslint-disable-next-line no-unused-vars
import * as React from 'react'
import { useState, useContext } from 'react'
import { useNavigate } from 'react-router-dom'
import OrderDetail from './OrderDetail'
import OrderInfoContext from '../../contexts/OrderInfoContext'
import booksStub from '../../components/booksStub'
import './Order.css'
import Navigation from '../../components/Navigation/Navigation'

function Order() {
    const [books, setBooks] = useState(booksStub())
    const { order, setOrder } = useContext(OrderInfoContext)

    const navigate = useNavigate()

    const totalPrice = books.reduce(
        (total, currentBook) =>
            total + currentBook.price * (currentBook.quantity - 1),
        0
    )

    const totalQuantity = books.reduce(
        (total, { quantity }) => total + (quantity - 1),
        0
    )

    const setQuantity = (id, quantity) => {
        setBooks(
            books.map((book) => (book.id !== id ? book : { ...book, quantity }))
        )
    }

    const handleReady = () => {
        setOrder({
            ...order,
            quantity: totalQuantity,
            price: totalPrice,
        })
        navigate('/order/user_order')
    }

    return (
        <>
            <Navigation />
            <OrderDetail books={books} setQuantity={setQuantity} />
            <button
                data-testid="button-ready"
                className="ButtonReady"
                type="button"
                onClick={handleReady}
            >
                Ready
            </button>
        </>
    )
}

export default Order
