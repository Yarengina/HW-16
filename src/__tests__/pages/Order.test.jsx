// eslint-disable-next-line no-unused-vars
import * as React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { render } from '../utils/renderWithRouter'
import Order from '../../pages/Order/Order'

const mockedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedNavigate,
}))

describe('<Order />', () => {
    it('should render successfully and match snapshot', () => {
        const { container } = render(<Order />)
        expect(container).toMatchSnapshot()
    })

    it('should have the button and navigate by clicking it', () => {
        render(<Order />)

        const button = screen.getByTestId('button-ready')
        expect(button).toBeInTheDocument()

        userEvent.click(button)
        expect(mockedNavigate).toHaveBeenCalledTimes(1)
    })
})
