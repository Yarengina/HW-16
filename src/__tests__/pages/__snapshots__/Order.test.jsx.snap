// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> should render successfully and match snapshot 1`] = `
<div>
  <nav
    class="LinkField"
  >
    <a
      aria-current="page"
      class="Link active"
      data-testid="login-link"
      href="/"
    >
      <span>
        Login
      </span>
    </a>
    <a
      class="Link"
      data-testid="about-link"
      href="/about"
    >
      <span>
        About
      </span>
    </a>
    <a
      class="Link"
      data-testid="order-link"
      href="/order"
    >
      <span>
        Order
      </span>
    </a>
  </nav>
  <div
    class="BookField"
  >
    <div
      class="Book"
      data-testid="todo-101"
    >
      <p
        class="Title"
      >
        Война и мир - Л.Н.Толстой
      </p>
      <p
        class="Price"
      >
        800
      </p>
      <a
        class="ReadMoreLink"
        href="/101"
      >
        Read more
      </a>
      <div
        class="Counter"
      >
        <button
          class="Buttons Plus"
          type="button"
        >
          Remove item
        </button>
        <p
          class="Quantity"
        >
          0
        </p>
        <button
          class="Buttons Minus"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="Book"
      data-testid="todo-102"
    >
      <p
        class="Title"
      >
        Две жизни - К.Е.Антарова
      </p>
      <p
        class="Price"
      >
        700
      </p>
      <a
        class="ReadMoreLink"
        href="/102"
      >
        Read more
      </a>
      <div
        class="Counter"
      >
        <button
          class="Buttons Plus"
          type="button"
        >
          Remove item
        </button>
        <p
          class="Quantity"
        >
          0
        </p>
        <button
          class="Buttons Minus"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="Book"
      data-testid="todo-103"
    >
      <p
        class="Title"
      >
        Разговор с богом - Н.Д.Уолша
      </p>
      <p
        class="Price"
      >
        1000
      </p>
      <a
        class="ReadMoreLink"
        href="/103"
      >
        Read more
      </a>
      <div
        class="Counter"
      >
        <button
          class="Buttons Plus"
          type="button"
        >
          Remove item
        </button>
        <p
          class="Quantity"
        >
          0
        </p>
        <button
          class="Buttons Minus"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="Book"
      data-testid="todo-104"
    >
      <p
        class="Title"
      >
        Хохот Шамана - В.П.Серкин
      </p>
      <p
        class="Price"
      >
        600
      </p>
      <a
        class="ReadMoreLink"
        href="/104"
      >
        Read more
      </a>
      <div
        class="Counter"
      >
        <button
          class="Buttons Plus"
          type="button"
        >
          Remove item
        </button>
        <p
          class="Quantity"
        >
          0
        </p>
        <button
          class="Buttons Minus"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
    <div
      class="Book"
      data-testid="todo-105"
    >
      <p
        class="Title"
      >
        Хроники Ехо - Макс Фрай
      </p>
      <p
        class="Price"
      >
        400
      </p>
      <a
        class="ReadMoreLink"
        href="/105"
      >
        Read more
      </a>
      <div
        class="Counter"
      >
        <button
          class="Buttons Plus"
          type="button"
        >
          Remove item
        </button>
        <p
          class="Quantity"
        >
          0
        </p>
        <button
          class="Buttons Minus"
          type="button"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
  <button
    class="ButtonReady"
    data-testid="button-ready"
    type="button"
  >
    Ready
  </button>
</div>
`;
